### ISSUE TITLE
InlineCompletion selectedSuggestInfo context enrichment

### ISSUE THREAD

**Issue by marrej** [microsoft/vscode#209814]
# Problem statement

TLDR: Having the snippetText exposed allows for deeper understanding of where the users cursor would land, allowing for more precise (and conditioned) inline completions. (with no completions added before the cursor position).

Currently the InlineCompletion context provides a limited view over the selectedSuggestInfo context (provides only parsed text and range of replacement). This makes it hard for the inline completion providers to better understand where the users cursor will end after accepting the suggest completion in case its a snippet text.

E.g. suggest completion `foo()` and suggest completion `foo($0)` may show totally different behavior when retrieving & applying the current completion. In current case (when `foo($0)` is used) the `foo()` as a whole would be used as a prefix (or none of the text would be used as prefix potentionally completing also between the `foo` & `()`), when instead the completion models might want to only complete from the cursor position, allowing to insert the arguments after `foo(` (e.g. `bar, baz)`).

# Solution

With the current selectedSuggestInfo, pass in optional semanticText, which would be provided inside context to the `provideInlineCompletions()` so that downstream providers can have tighter control over the inline completion.

(The extension of the selectedSuggestInfo could be added to the current vscode.proposed.inlineCompletionsAdditions.d.ts)
Happy to add the change, if agreed on.

**Comment by hediet** [comment]
I understand the problem statement, but I don't understand the relevance of the problem.

Can you motivate why this problem should be fixed? What is the common use case that users would run into?

I don't like about the solution that it breaks abstraction. `semanticText: string` is not really a semantic data type.
So far, I tried to hide suggest implementation details from the inline completion. Augmenting snippets was more of an after-thought and not the primary use-case for completion-augmentation.

**Comment by marrej** [comment]
Hi! Sorry for the late response, this slipped my radar due to load of other priorities.

I think I should have started with saying, **that we are modifying the `augment` function**, to actually return always only the text before the the first snippet, because we found this to be too restrictive (dropping significant amount of completions). So as a leading point to this one, I would say it would be great to allow completions to augment, not only based on full prefix alignment, but also to check based on the snippet params, whether the completions are aligned.

But to answer your questions:

> I don't like about the proposed solution that it breaks abstraction. 

I agree that it kinda breaks the abstraction, but truth be told, both of these components are interconnected. And having just the pure parsed text of the semantic completion a bit misrepresents what actually flows from LSP.
The common use case very much depends on the completions that flow from the LSP server. If in most cases they contain only the parameter/function names, without additional context, the inline suggest wouldn't be impacted.

> Can you motivate why this problem should be fixed? What is the common use case that users would run into?

But take for example a case when the LSP server returns `foo($0)`. Potentionally this means that there is content to be inserted in space of the $0, but the core processing logic will return only `foo()`.
The same would happen also in case when the server returns `baz()` (note that in this case the baz would not have any arguments, and would look like `const baz = () => {}`  unlike previous `const foo = (arg1, arg2) => {} `).

So there are 2 very different completions, which may require different treatment, but won't get it. Neither the code completion provider would know whether it can insert any parameters inside the function so its not easily how much impact is lost, but as well currently in the core functionality also it wouldn't be possible to attach those parameters.

**So by providing the snippets to the provider** we can have 
- a better understanding of the completions provided
- Increase the amount of (successful/accepted) completions
- Increase the user happiness as they may get unblocked easier & more e.g. when typing out function parameters.

**Comment by hediet** [comment]
So you want to use `${i}` in a snippet as signal for the AI to fill in the "placeholder"?

**Comment by marrej** [comment]
> It sounds like it could work, but I'm not sure it should be solved like this. Instead, the AI should get this data from LSP server directly somehow.

We can either use that, or we can provide an array of split tokens, so e.g. for `foo(${1}, ${2})`, we could either provide the direct string, or `['foo(', ', ', ')']` which extensions/users can read as between these tokens anything can be inserted and the first token being the actual prefix.

I wouldn't want to delegate this data to the LSP server, as that: 

- introduces unneccessary potentional concurency issues
- requires all AI extensions that vscode has to connect to the LSP (while some connect directly to local models, e.g. `Tabby`)
- Complicates the whole interaction for the extensions & potentially increases latency, which could have this information made ready
